runtime/src/bank.rs

```diff
@@ -4138,28 +4160,37 @@ impl Bank {
             .is_hash_valid_for_age(hash, max_age)
     }
 
-    fn check_message_for_nonce(&self, message: &SanitizedMessage) -> Option<TransactionAccount> {
-        message
+    fn check_message_for_nonce(
+        &self,
+        message: &SanitizedMessage,
+    ) -> std::result::Result<TransactionAccount, NonceError> {
+        let nonce_address = message
             .get_durable_nonce(self.feature_set.is_active(&nonce_must_be_writable::id()))
-            .and_then(|nonce_address| {
-                self.get_account_with_fixed_root(nonce_address)
-                    .map(|nonce_account| (*nonce_address, nonce_account))
-            })
-            .filter(|(_, nonce_account)| {
-                nonce_account::verify_nonce_account(nonce_account, message.recent_blockhash())
-            })
+            .ok_or(NonceError::NonceAddressNotFound)?;
+        let nonce_account = self
+            .get_account_with_fixed_root(nonce_address)
+            .ok_or(NonceError::NonceAccountNotFound)?;
+        if nonce_account::verify_nonce_account(&nonce_account, message.recent_blockhash()) {
```

Comment:
I was considering plumbing `NonceError` through here too, so we can discern between an invalid nonce account and a blockhash mismatch, but I think it'd just be informational (and we'd eat that info immediately)